src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -874,6 +898,10 @@ private Query filterQueryIfNeeded(Query query, String[] types) {
 
     public static final String INDEX_REFRESH_INTERVAL = "index.refresh_interval";
 
+    public void addFailedEngineListener(Engine.FailedEngineListener failedEngineListener) {
+        this.failedEngineListener.failedEngineListeners.add(failedEngineListener);
```

Comment:
Waaaay too many "failedEngineListener"s in this line :) maybe we can rename one/some to make it a little less mind-boggling?